src/components/TopBar/createMenu.jsx

```diff
@@ -0,0 +1,107 @@
+/* eslint-disable react/no-multi-comp */
+
+import classnames from 'classnames';
+import {connect} from 'react-redux';
+import constant from 'lodash/constant';
+import onClickOutside from 'react-onclickoutside';
+import preventClickthrough from 'react-prevent-clickthrough';
+import property from 'lodash/property';
+import PropTypes from 'prop-types';
+import React from 'react';
+import {closeTopBarMenu, toggleTopBarMenu} from '../../actions';
+import {getOpenTopBarMenu} from '../../selectors';
+
+export function MenuItem({children, isEnabled, onClick}) {
+  return (
+    <div
+      className={classnames('top-bar__menu-item',
+        {'top-bar__menu-item_active': isEnabled},
+      )}
+      onClick={onClick}
+    >
+      {children}
+    </div>
+  );
+}
+
+MenuItem.propTypes = {
+  children: PropTypes.node.isRequired,
+  isEnabled: PropTypes.bool,
+  onClick: PropTypes.func.isRequired,
+};
+
+MenuItem.defaultProps = {
+  isEnabled: false,
+};
+
+export default function createMenu({
+  isVisible = constant(true),
+  renderItems,
+  name,
+}) {
+  function mapStateToProps(state) {
+    const isOpen = getOpenTopBarMenu(state) === name;
+    return {
+      isOpen,
+      disableOnClickOutside: !isOpen,
+    };
+  }
+
+  function mapDispatchToProps(dispatch) {
+    return {
+      onClose() {
+        dispatch(closeTopBarMenu(name));
+      },
+
+      onToggle() {
+        dispatch(toggleTopBarMenu(name));
+      },
+    };
+  }
+
+  return function createMenuWithMappedProps(Label) {
```

Comment:
A name like `LaunchButton` or `MenuLaunchButton` or `MenuLauncher` might be a little more helpful for figuring out what this argument is for (assuming that I've got it figured out 😑).